commit e88f7376fe68dbf4ebaf11fad1513ce700b45860
Author: Sarah Boyce <42296566+sarahboyce@users.noreply.github.com>
Date:   Tue Feb 25 09:40:54 2025 +0100

    [4.2.x] Fixed CVE-2025-26699 -- Mitigated potential DoS in wordwrap template filter.
    
    Thanks sw0rd1ight for the report.
    
    Backport of 55d89e25f4115c5674cdd9b9bcba2bb2bb6d820b from main.

diff --git a/django/utils/text.py b/django/utils/text.py
index e1b835e0e2..81ae88dc76 100644
--- a/django/utils/text.py
+++ b/django/utils/text.py
@@ -1,6 +1,7 @@
 import gzip
 import re
 import secrets
+import textwrap
 import unicodedata
 from gzip import GzipFile
 from gzip import compress as gzip_compress
@@ -87,34 +88,25 @@ re_camel_case = _lazy_re_compile(r"(((?<=[a-z])[A-Z])|([A-Z](?![A-Z]|$)))")
 @keep_lazy_text
 def wrap(text, width):
     """
     A word-wrap function that preserves existing line breaks. Expects that
     existing line breaks are posix newlines.
 
     Preserve all white space except added line breaks consume the space on
     which they break the line.
 
     Don't wrap long words, thus the output text may have lines longer than
     ``width``.
     """
 
-    def _generator():
-        for line in text.splitlines(True):  # True keeps trailing linebreaks
-            max_width = min((line.endswith("\n") and width + 1 or width), width)
-            while len(line) > max_width:
-                space = line[: max_width + 1].rfind(" ") + 1
-                if space == 0:
-                    space = line.find(" ") + 1
-                    if space == 0:
-                        yield line
-                        line = ""
-                        break
-                yield "%s\n" % line[: space - 1]
-                line = line[space:]
-                max_width = min((line.endswith("\n") and width + 1 or width), width)
-            if line:
-                yield line
-
-    return "".join(_generator())
+    wrapper = textwrap.TextWrapper(
+        width=width,
+        break_long_words=False,
+        break_on_hyphens=False,
+    )
+    result = []
+    for line in text.splitlines(True):
+        result.extend(wrapper.wrap(line))
+    return "\n".join(result)
 
 
 class Truncator(SimpleLazyObject):
diff --git a/docs/releases/4.2.20.txt b/docs/releases/4.2.20.txt
index c71fa05f43..5849fe2a42 100644
--- a/docs/releases/4.2.20.txt
+++ b/docs/releases/4.2.20.txt
@@ -5,3 +5,9 @@ Django 4.2.20 release notes
 *March 6, 2025*
 
 Django 4.2.20 fixes a security issue with severity "moderate" in 4.2.19.
+
+CVE-2025-26699: Potential denial-of-service vulnerability in ``django.utils.text.wrap()``
+=========================================================================================
+
+The ``wrap()`` and :tfilter:`wordwrap` template filter were subject to a
+potential denial-of-service attack when used with very long strings.
diff --git a/tests/template_tests/filter_tests/test_wordwrap.py b/tests/template_tests/filter_tests/test_wordwrap.py
index 88fbd274da..4afa1dd234 100644
--- a/tests/template_tests/filter_tests/test_wordwrap.py
+++ b/tests/template_tests/filter_tests/test_wordwrap.py
@@ -32,49 +32,60 @@ class WordwrapTests(SimpleTestCase):
 class FunctionTests(SimpleTestCase):
     def test_wrap(self):
         self.assertEqual(
             wordwrap(
                 "this is a long paragraph of text that really needs to be wrapped I'm "
                 "afraid",
                 14,
             ),
             "this is a long\nparagraph of\ntext that\nreally needs\nto be wrapped\n"
             "I'm afraid",
         )
 
     def test_indent(self):
         self.assertEqual(
             wordwrap(
                 "this is a short paragraph of text.\n  But this line should be "
                 "indented",
                 14,
             ),
             "this is a\nshort\nparagraph of\ntext.\n  But this\nline should be\n"
             "indented",
         )
 
     def test_indent2(self):
         self.assertEqual(
             wordwrap(
                 "this is a short paragraph of text.\n  But this line should be "
                 "indented",
                 15,
             ),
             "this is a short\nparagraph of\ntext.\n  But this line\nshould be\n"
             "indented",
         )
 
     def test_non_string_input(self):
         self.assertEqual(wordwrap(123, 2), "123")
 
     def test_wrap_lazy_string(self):
         self.assertEqual(
             wordwrap(
                 lazystr(
                     "this is a long paragraph of text that really needs to be wrapped "
                     "I'm afraid"
                 ),
                 14,
             ),
             "this is a long\nparagraph of\ntext that\nreally needs\nto be wrapped\n"
             "I'm afraid",
         )
+
+    def test_wrap_long_text(self):
+        long_text = (
+            "this is a long paragraph of text that really needs"
+            " to be wrapped I'm afraid " * 20_000
+        )
+        self.assertIn(
+            "this is a\nlong\nparagraph\nof text\nthat\nreally\nneeds to\nbe wrapped\n"
+            "I'm afraid",
+            wordwrap(long_text, 10),
+        )
